### ISSUE TITLE
PyPI prerelease version badge

### ISSUE THREAD

**Issue by paulmelnikow** [badges/shields#4692]
:clipboard: **Description**

I just published a prerelease of tinyobjloader:

https://pypi.org/project/tinyobjloader/#history

As far as I can tell, there isn't a badge for displaying the latest prerelease version.

It looks like if there are _only_ prereleases, they do show correctly, e.g.

https://img.shields.io/pypi/v/black
![](https://img.shields.io/pypi/v/black)

:link: **Data**

It looks like the data source we use now includes both of the releases for this project, 0.1 and 2.0.0rc5.

https://pypi.org/pypi/tinyobjloader/json

:microphone: **Motivation**

It's common for projects to run prereleases for a long time, and it's useful to show users the latest published prerelease.

**Comment by chris48s** [comment]
> As far as I can tell, there isn't a badge for displaying the latest prerelease version

I think the reason we haven't done this is: With PyPI, we defer to the registry to tell us which is the "latest" version. Beyond that, PyPI's API doesn't explicitly mark releases as pre-release or stable. This means we would have to make our own decision about which releases are pre-release and which are stable.

Most (all?) packages on PyPI follow PEP440 rather than semver, so we'd need a javascript implementation of PEP440 for that case (which is a thing that does exist https://github.com/renovatebot/pep440 ). I'm not 100% sure if PyPI _enforces_ PEP440 or not though. If it does, "all package versions are PEP440-compliant" is a reasonable assumption and we could just roll with that. If it doesn't, we're then on shaky ground with that assumption as there will inevitably be some packages in the registry with non-compliant versions.

**Comment by paulmelnikow** [comment]
I don’t know if PyPI enforces PEP440 either. It does seem to normalize version numbers (e.g. if you upload a package with `beta6` it will allow fetching it with `beta6` but at least lists it as `b6`: https://pypi.org/project/polliwog/#history

> Beyond that, PyPI's API doesn't explicitly mark releases as pre-release or stable.

Prereleases are defined in PEP440 and fortunately would be relatively easy to detect using regex.

One option that would solve my use case would be to have an option that returns the latest release by date. There are upload dates in the JSON, although they might have to be aggregated across the various sdists and wheels.

I know you worked a lot on the GitHub release badges so I bet you have this problem a lot more in your head than I do. 🤔

**Comment by chris48s** [comment]
Just thinking about it in slightly more detail..

PyPI does flag releases as pre-release in the front-end (its just not exposed via the API):

![Screenshot at 2020-02-27 17-20-52](https://user-images.githubusercontent.com/6025893/75468750-a0b59a80-5985-11ea-9ce0-1ce9d8e7cc46.png)

Looking at the warehouse source, the code for doing that is:

https://github.com/pypa/warehouse/blob/1fbb4ac752e68b5840b9e09b68e44a165569bfa6/warehouse/migrations/versions/e7b09b5c089d_add_pep440_is_prerelease.py#L29-L34

so we could replicate the same method for identifying whether a version is stable or pre-release but I think we'd still want to use https://github.com/renovatebot/pep440 for comparing/sorting versions to determine the latest in the pre-release case.

**Comment by huan** [comment]
It would be great if we can get a Pre Release PyPI version badge just like NPM `@ next`. +1

**Comment by graingert** [comment]
>  latest in the pre-release case.

latest by release date or greatest version number?

**Comment by graingert** [comment]
Here's some discussion around making an API https://github.com/pypa/warehouse/issues/4663

**Comment by chris48s** [comment]
Thanks :+1:  I've subscribed to https://github.com/pypa/warehouse/pull/8615 - we'll see how it plays out

**Comment by Kodiologist** [comment]
Looks like that won't happen soon.

> One option that would solve my use case would be to have an option that returns the latest release by date.

This is probably the way to go, and would work well for the project I'm thinking of right now, https://github.com/hylang/hy.

**Comment by Bibo-Joshi** [comment]
Another option would be to add a parameter to the badge that allows to specify a specific version (or add another endpoint for that). E.g. `/pypi/pyversions/:packageName?version=v0.0a1`.

**Comment by EwoutH** [comment]
The [NuGet Version](https://shields.io/badges/nu-get-version) badge supports pre-releases.

![image](https://github.com/badges/shields/assets/15776622/53c10273-6c2b-43f9-b006-f7a6c86aa194)

It would be really nice if there was a similar option for PyPI.

**Comment by natke** [comment]
+1 to the pre option. If you have both Python and NuGet packages, your badges and want to show the latest, your badges look a little weird

**Comment by EwoutH** [comment]
I again encountered a project where it would be very beneficial to have a pre-release version badge.

**Comment by EwoutH** [comment]
I don't think this is that difficult. Updating [pypi-helpers.js](https://github.com/badges/shields/blob/master/services/pypi/pypi-helpers.js) would work parsing pre-release versions, right?

```js
function parsePypiVersionString(str) {
  if (typeof str !== 'string') {
    return false;
  }
  
  const versionPattern = /^(?<major>\d+)\.(?<minor>\d+)(?:\.(?<patch>\d+))?(?<prerelease>[a-z]+\d+)?$/;
  const match = str.match(versionPattern);

  if (!match) {
    return false;
  }

  const { major, minor, patch = 0, prerelease = '' } = match.groups;
  
  return {
    major: parseInt(major),
    minor: parseInt(minor),
    patch: parseInt(patch),
    prerelease,
  };
}

// Sort an array of PyPI versions low to high.
function sortPypiVersions(versions) {
  return versions.sort((a, b) => {
    const versionA = parsePypiVersionString(a);
    const versionB = parsePypiVersionString(b);

    if (versionA.major !== versionB.major) {
      return versionA.major - versionB.major;
    }

    if (versionA.minor !== versionB.minor) {
      return versionA.minor - versionB.minor;
    }

    if (versionA.patch !== versionB.patch) {
      return versionA.patch - versionB.patch;
    }

    // Handle pre-release versions (a0 < b0 < rc0)
    if (versionA.prerelease && versionB.prerelease) {
      return versionA.prerelease.localeCompare(versionB.prerelease);
    }

    // If one version is pre-release and the other is not, the pre-release version is lower
    if (versionA.prerelease) {
      return -1;
    }

    if (versionB.prerelease) {
      return 1;
    }

    // Versions are identical
    return 0;
  });
}

// Example usage:
const versions = ['1.11', '1.11a1', '1.11b2', '1.11rc1', '1.12', '2.0'];
const sortedVersions = sortPypiVersions(versions);
console.log(sortedVersions); // ['1.11a1', '1.11b2', '1.11rc1', '1.11', '1.12', '2.0']

export {
  parseClassifiers,
  parsePypiVersionString,
  sortPypiVersions,
  getLicenses,
  getPackageFormats,
}

```

**Comment by HinTak** [comment]
Ping - actually we like the pre-release versions to have some visibility so that they get some testing before the next release. Have had over 10 pre-releases now, and will likely have another 10...

**Comment by rayrayraykk** [comment]
Looking forward to this feature
